### ISSUE TITLE
add support for requiring an extra factor (PIN, passphrase or pattern) with fingerprint unlock to replace split boot / lockscreen password

### ISSUE THREAD

**Issue by thestinger** [GrapheneOS/os_issue_tracker#28]

**Comment by jengo9332** [comment]
@thestinger has the split boot/lockscreen password been migrated to GrapheneOS?

**Comment by thestinger** [comment]
That feature is obsolete and isn't applicable to modern Android. It hasn't existed for years. There's no such thing as a boot password on devices with the modern encryption format introduced with 1st generation Pixels. There are per-profile encryption keys, not a global encryption key and they can become at rest again when profiles are logged out. Fingerprint unlock is an existing implementation of a secondary unlock mechanism available after the initial unlock until a timeout expires. This issue is about extending it with support for a PIN as a far superior successor to the obsolete approach. It's not implemented which is why it's an open issue.

**Comment by thestinger** [comment]
That feature is obsolete and isn't applicable to modern Android. It hasn't existed for years. There's no such thing as a boot password on devices with the modern encryption format introduced with 1st generation Pixels. There are per-profile encryption keys, not a global encryption key and they can become at rest again when profiles are logged out. Fingerprint unlock is an existing implementation of a secondary unlock mechanism available after the initial unlock until a timeout expires. This issue is about extending it with support for a PIN as a far superior successor to the obsolete approach. It's not implemented which is why it's an open issue. This is the replacement for the obsolete feature, and will have strictly better security properties along with significantly better usability.

**Comment by thestinger** [comment]
That feature is obsolete and isn't applicable to modern Android. It hasn't existed for years. There's no such thing as a boot password on devices with the modern encryption format introduced with 1st generation Pixels. There are per-profile encryption keys, not a global encryption key and they can become at rest again when profiles are logged out. Fingerprint unlock is an existing implementation of a secondary unlock mechanism available after the initial unlock until a timeout expires. This issue is about extending it with support for a PIN as a far superior successor to the obsolete approach. It's not implemented which is why it's an open issue. This is the replacement for the obsolete feature, and will have strictly better security properties along with significantly better usability.

The primary unlock mechanism (which should be set to a strong passphrase) is used to derive the main encryption key for the profile and is always required for the initial unlock of the profile. The secondary unlock mechanism can then be used until it times out. Both the primary and secondary unlock mechanism are used to protect the keystore. Keys can be set to be usable only when the profile is unlocked in order to keep them at rest when it's locked:

https://developer.android.com/reference/android/security/keystore/KeyProtection.Builder.html#setUnlockedDeviceRequired(boolean)

I hope this makes sense. The legacy feature you're talking about cannot exist anymore and this issue is about implementing a superior replacement that fits into the new system.

**Comment by throwException** [comment]
Might I (and hopefully others) put out a bounty to motivate some developer to implement this?

(I have no experience with such bounties yet)

**Comment by lepras** [comment]
after some searching I found these:

[Can I make Android require Fingerprint + PIN to unlock the screen? | StackExchange](https://android.stackexchange.com/questions/139653/can-i-make-android-require-fingerprint-pin-to-unlock-the-screen)

[Lockscreen and authentication improvements in Android 11 | GoogleBlog](https://android-developers.googleblog.com/2020/09/lockscreen-and-authentication.html)

[Better Biometrics in Android P | Google Blog](https://android-developers.googleblog.com/2018/06/better-biometrics-in-android-p.html)

[One Biometric API Over all Android | googleBlog](https://android-developers.googleblog.com/2019/10/one-biometric-api-over-all-android.html)

will dig deeper and find some API's that we can use and come up with a design.

**Comment by Foxtrek64** [comment]
A bit of an implementation question. What second factor options are we looking at allowing here - just pins? Or would we allow passphrases or patterns as well?

Primary security is through the fingerprint, so I imagine the security of the secondary factor would not be *as* important as the first factor.

**Comment by throwException** [comment]
Yes, a PIN makes perfect sense, if this quick method is locked out after only a few tries.

The number of tries should be configurable as it heavily depends on the use case as well as the length of the PIN.

The lockout should be lifted as soon as the primary passphrase unlock is entered correctly.

There should be the option to scramble the PIN entry layout.

**Comment by Foxtrek64** [comment]
Thanks for the info! I look forward to seeing a PR on this.

**Comment by schrauger** [comment]
> after some searching I found these:
> 
> [Can I make Android require Fingerprint + PIN to unlock the screen? | StackExchange](https://android.stackexchange.com/questions/139653/can-i-make-android-require-fingerprint-pin-to-unlock-the-screen)

I remember that question! At the time, I came up with a somewhat convoluted and not-exactly-secure solution using Tasker. I'm really looking forward to this upcoming feature in grapheneos. It's one that I have been wanting for nearly a decade!

**Comment by nyaasi** [comment]
Just to make sure, but the duress pin is still functional with the intended 2nd factor and will wipe the device, right?

**Comment by richardebeling** [comment]
Note that some users are asking for the PIN as an _alternative_ to the fingerprint, not as an additional factor
* In #3795 and [on the discussion forum](https://discuss.grapheneos.org/d/15636-setting-second-shorter-pin-for-afu-unlocks), users asked for a long password for encryption and a short pin for AFU-unlocking, without using fingerprints at all
* In #524, a user asked about the option to use the short pin instead of the fingerprint for unlocking. I would also like to have that feature, for situations where fingerprint detection stops working for whatever reason (sweaty/dirty hands, injury, fingerprint sensor broken, ...)

These require more degrees of freedom for the user to configure combined authentication methods than just adding a second PIN-factor to fingerprint-AFU-unlocking. It would be nice if this issue could reflect that.

**Comment by secretmango** [comment]
Yes, fingerprint + pin/password is the best combo

Fingeprint is secure but can be forced

Password cannot just be forced but can be spied on with CCTV cameras etc.

**Comment by egomaniac9023** [comment]
How is the progress here? Where I can view the updates related to it (code changes) ?
